### PR TITLE
Add Three Tier Subscriber+ subscription path to support-workers

### PR DESCRIPTION
**Pull request by jamesgorrie**
Depends on
- #5628 
- #5627

## What are you doing in this PR?
Adds a route to the `support-workers` step functions that
- given the right conditions
  - `state.acquisitionData.threeTierCreateSupporterPlusSubscription == true` (set [here](https://github.com/guardian/support-frontend/pull/5628))
  - `state.product.productType == GuardiaWeekly`
- Replicate the subscription
- Replace the `GuardianWeekly` with the relevant `SupporterPlus` product
  - This is based on `currency` and `Monthly | Annual`

Screenshot 2024-01-23 at 10 25 47

https://link.excalidraw.com/readonly/04tDyUsXkGa7qBfnDnWo

[**Trello Card**](https://trello.com/c/zh96UwvO/486-create-support-subscription-for-supporter-buying-weekly-from-the-third-tier-in-3-tier-test)

## Why are you doing this?
This will allow us to test the Three Tier setup by allowing a person to order the Weekly subscription, which we then bolt the Supporter+ product onto as a secondary step.

## How to test

- Deploy to `CODE`
- Try order weekly with the corresponding query string parameters
- Hope I get a Weekly and Supporter+ subscription

## How can we measure success?

We can start running the test.

### Code

The code is currently all written in `YAML` / `Handlebars` / `CloudFormation`. All the values for the test are also hardcoded within that.

This isn't the most scalable way to do this and we will have duplication with other hard coded areas of the site (this is going to be hardcoded in the frontend too). It is however very easy to change and also touches as little of the existing code and infrastructure as I could possibly work out how, which is good for a test.

### Thank you pages AKA a successful purchase

This will send a success message to the person as soon as their Weekly subscription has past. It will not alert them to any failures in their subscription to the Supporter+ flow. 

This _should be_ OK. If your payment has worked for Weekly, there is little chance it won't for the Support+. We might want to think about putting extra alerting in that Customer Support might be able to respond to though.

## Next steps

- Add promo codes

### CODE REVIEW

**Comment by jamesgorrie** [comment]
We don't ever send [this data](https://github.com/guardian/support-frontend/pull/5629/files#diff-9059d2bc2c21a4a3496fd1bd1a80b1a928fd9e39418d4b87ed444e56f8a1a732R91-R94) from the client so we should be good to merge to PROD as V1.

File: support-workers/cloud-formation/src/templates/state-machine.yaml [hunk 1]
```diff
@@ -68,6 +68,93 @@ States:
...
+        subscriptionNumber.$: $.state.sendThankYouEmailState.subscriptionNumber
+        firstDeliveryDate.$: $.state.sendThankYouEmailState.firstDeliveryDate
+        productType.$: $.state.sendThankYouEmailState.productType
+        requestId.$: $.state.requestId
+        analyticsInfo.$: $.state.analyticsInfo
+        acquisitionData.$: $.state.acquisitionData
+      error.$: $.error
+      requestInfo.$: $.requestInfo
+    Next: ThreeTierChooseSubscriberPlusProduct
+
```
Comment (jamesgorrie) [inline support-workers/cloud-formation/src/templates/state-machine.yaml:126]: Add a note here how the decision is being made.

File: support-workers/cloud-formation/src/templates/state-machine.yaml [hunk 2]
```diff
@@ -68,6 +68,93 @@ States:
...
+    Next: ThreeTierShouldCreateSupporterPlusSubscription
+
+  ThreeTierShouldCreateSupporterPlusSubscription:
+    Type: Choice
+    Choices:
+      - Next: ThreeTierUnpackSendThankYouEmailState
+        And:
+          - Variable: $.state.acquisitionData.threeTierCreateSupporterPlusSubscription
+            IsPresent: true
+          - Variable: $.state.acquisitionData.threeTierCreateSupporterPlusSubscription
```
Comment (jamesgorrie) [inline support-workers/cloud-formation/src/templates/state-machine.yaml:93]: This is the part that should avoid us having this run as we never actually send this state down the pipes as yet.

Comment (jamesgorrie) [reply support-workers/cloud-formation/src/templates/state-machine.yaml:93]: Even with the `threeTierCreateSupporterPlusSubscription=true` param

Screenshot 2024-01-24 at 11 47 09

The Pipeline skips the three tier branch:
Screenshot 2024-01-24 at 11 51 50

File: support-workers/cloud-formation/src/templates/state-machine.yaml (hunk 1 above)
Comment (jamesgorrie) [reply support-workers/cloud-formation/src/templates/state-machine.yaml:126]: [Done](https://github.com/guardian/support-frontend/pull/5629/commits/c69aed8084e09f8f80b6a1e3569534556f7e473f)

File: support-workers/cloud-formation/src/build-cfn.js [hunk 3]
```diff
@@ -10,7 +10,9 @@ try { fs.mkdirSync('../target'); } catch(err) { }
 let partials = {
   retry: loadTemplate('retries.yaml'),
   emailRetry: loadTemplate('emailRetries.yaml'),
-  catch: loadTemplate('catch.yaml')
+  catch: loadTemplate('catch.yaml'),
+  subscriberPlusProduct: loadTemplate('subscriberPlusProduct.yaml'),
```
Comment (rupertbates) [inline support-workers/cloud-formation/src/build-cfn.js:14]:
```suggestion
  supporterPlusProduct: loadTemplate('subscriberPlusProduct.yaml'),
```

File: support-workers/cloud-formation/src/build-cfn.js [hunk 4]
```diff
@@ -10,7 +10,9 @@ try { fs.mkdirSync('../target'); } catch(err) { }
...
+  subscriberPlusProductChoice: loadTemplate('subscriberPlusProductChoice.yaml')
```
Comment (rupertbates) [inline support-workers/cloud-formation/src/build-cfn.js:15]:
```suggestion
  supporterPlusProductChoice: loadTemplate('supporterPlusProductChoice.yaml')
```

File: support-workers/cloud-formation/src/templates/state-machine.yaml [hunk 5]
```diff
@@ -68,6 +68,95 @@ States:
...
+        firstDeliveryDate.$: $.state.sendThankYouEmailState.firstDeliveryDate
+        productType.$: $.state.sendThankYouEmailState.productType
+        requestId.$: $.state.requestId
+        analyticsInfo.$: $.state.analyticsInfo
+        acquisitionData.$: $.state.acquisitionData
+      error.$: $.error
+      requestInfo.$: $.requestInfo
+    Next: ThreeTierChooseSubscriberPlusProduct
+
+  ThreeTierChooseSubscriberPlusProduct:
```
Comment (rupertbates) [inline support-workers/cloud-formation/src/templates/state-machine.yaml:127]:
```suggestion
  ThreeTierChooseSupporterPlusProduct:
```

File: support-workers/cloud-formation/src/templates/subscriberPlusProduct.yaml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# This step ID is called from and should match ./subscriberPlusProductChoice.yaml
+ThreeTierSubscriberPlus{{billingPeriod}}{{currency}}:
```
Comment (rupertbates) [inline support-workers/cloud-formation/src/templates/subscriberPlusProduct.yaml:2]:
```suggestion
ThreeTierSupporterPlus{{billingPeriod}}{{currency}}:
```

File: support-workers/cloud-formation/src/templates/subscriberPlusProductChoice.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# This step ID is generated and should match in ./subscriberPlusProduct.yaml
```
Comment (rupertbates) [inline support-workers/cloud-formation/src/templates/subscriberPlusProductChoice.yaml:1]:
```suggestion
# This step ID is generated and should match in ./supporterPlusProduct.yaml
```

Plus filename

**Comment by jamesgorrie** [comment]
@rupertbates I have no idea why "subscriber" is so embedded in my psyche.

**Review by rupertbates — APPROVED** [review]
Looks good

**Pull request merged**